binderhub/repoproviders.py

```diff
@@ -186,16 +187,32 @@ class GitRepoProvider(RepoProvider):
 
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        url, resolved_ref = self.spec.rsplit('/', 1)
+        url, unresolved_ref = self.spec.split('/', 1)
         self.repo = urllib.parse.unquote(url)
-        if not resolved_ref:
-            raise ValueError("`resolved_ref` must be specified as a query parameter for the basic git provider")
-        self.sha1_validate(resolved_ref)
-        self.resolved_ref = resolved_ref
-        self.unresolved_ref = resolved_ref
+        self.unresolved_ref = urllib.parse.unquote(unresolved_ref)
+        if not self.unresolved_ref:
+            raise ValueError("`unresolved_ref` must be specified as a query parameter for the basic git provider")
+        try:
+            self.sha1_validate(unresolved_ref)
+            self.resolved_ref = unresolved_ref
+        except ValueError:
+            pass
```

Comment:
There are a few things here which aren't obvious from reading it a few times so maybe we can add some comments.

Why are we now swallowing exceptions here?

On L196 calling `self.sha1_validate(unresolved_ref)` should fail a lot because users can pass in a value like `master` so why call it now and not after we resolve the reference?

It seems weird to set `self.resolved_ref` to the value of `unresolved_ref`. I think we should delay that to when we have a resolved ref (and it it to `None` in the meantime).

Reply:
We pass on an error since we postpone resolving the ref to the getter of `resolved_ref` if the ref is not a commit SHA. It makes sense to move this sha1 check to the getter for the `resolved_ref` and not set the `resolved_ref` at all in the constructor. This is in line with the other repo providers.

Reply:
I also appreciate managing all logic within the getter as there is one. It was confusing to me the first time I read this through that some of it was found there but not all of it.

Reply:
Is the current implementation satisfactory? 

The only thing I'm not 100% sure about is setting the `resolved_ref` in the `else` block. It is a good practice to only have code in the `try` block that produces the error you want to catch. However, it could be more readable if the `resolved_ref` is set directly after validating it in the `try` block. I tried to make it more readable by adding some comments, but still assign it in the `else` block.